Improving progress logging and memory performance

cr opened the pull request:
* Switching regression, scan, and srcupdate to chunked operation
* Revised progress logging to work across chunks
* Removing default source limit
* Replacing progress_bar with progress_logger
* Stripping progress tracking logic from workers
* Fixing issue where workers would sometimes hang after ctrl-c during worker startup
* Improved ctrl-c shutdown behavior, making it faster
* Terminology: replacing URLs with hosts
* Improving stability with truncated log files
* Adding ProgressLogger tests
* Version bump to 3.2.0a1

cr commented:
This will fix #44.

cr commented:
This is also fixing some #67.

mwobensmith commented:
I am running a full top sites regression pass on this before I can approve, but once I finish, I'm pretty confident that we can land this.

mwobensmith commented:
Running the full list, hitting some problems. It has hung for a few hours, with the debug output below (edited for length). 

So, I'm concerned about this. Can you try a full regression run?

```

2017-07-28 00:17:44 INFO Progress tlscanary.progress 35% 190558/544372, 15% overhead, -- hosts/s net, -- hosts/s gross, ETA --
2017-07-28 00:22:44 INFO Progress tlscanary.progress 35% 190558/544372, 15% overhead, -- hosts/s net, -- hosts/s gross, ETA --
2017-07-28 00:27:45 INFO Progress tlscanary.progress 35% 190558/544372, 15% overhead, -- hosts/s net, -- hosts/s gross, ETA --
2017-07-28 00:32:45 INFO Progress tlscanary.progress 35% 190558/544372, 15% overhead, -- hosts/s net, -- hosts/s gross, ETA --
2017-07-28 00:37:45 INFO Progress tlscanary.progress 35% 190558/544372, 15% overhead, -- hosts/s net, -- hosts/s gross, ETA --

(snip)

2017-07-28 03:27:54 INFO Progress tlscanary.progress 35% 190558/544372, 15% overhead, -- hosts/s net, -- hosts/s gross, ETA --
2017-07-28 03:32:54 INFO Progress tlscanary.progress 35% 190558/544372, 15% overhead, -- hosts/s net, -- hosts/s gross, ETA --
2017-07-28 03:37:54 INFO Progress tlscanary.progress 35% 190558/544372, 15% overhead, -- hosts/s net, -- hosts/s gross, ETA --
2017-07-28 03:42:55 INFO Progress tlscanary.progress 35% 190558/544372, 15% overhead, -- hosts/s net, -- hosts/s gross, ETA --
2017-07-28 03:47:55 INFO Progress tlscanary.progress 35% 190558/544372, 15% overhead, -- hosts/s net, -- hosts/s gross, ETA --
2017-07-28 03:52:55 INFO Progress tlscanary.progress 35% 190558/544372, 15% overhead, -- hosts/s net, -- hosts/s gross, ETA --
2017-07-28 03:57:55 INFO Progress tlscanary.progress 35% 190558/544372, 15% overhead, -- hosts/s net, -- hosts/s gross, ETA --
2017-07-28 04:02:56 INFO Progress tlscanary.progress 35% 190558/544372, 15% overhead, -- hosts/s net, -- hosts/s gross, ETA --

```

mwobensmith commented:
Eventually cancelled via ctrl-C, here is stdout:
```

2017-07-28 12:13:21 INFO Progress tlscanary.progress 35% 190558/544372, 15% overhead, -- hosts/s net, -- hosts/s gross, ETA --
2017-07-28 12:18:21 INFO Progress tlscanary.progress 35% 190558/544372, 15% overhead, -- hosts/s net, -- hosts/s gross, ETA --
2017-07-28 12:23:21 INFO Progress tlscanary.progress 35% 190558/544372, 15% overhead, -- hosts/s net, -- hosts/s gross, ETA --
2017-07-28 12:28:22 INFO Progress tlscanary.progress 35% 190558/544372, 15% overhead, -- hosts/s net, -- hosts/s gross, ETA --
^[[B^[[B^[[B^[[B^[[B^[[B^C
2017-07-28 12:29:46 CRITICAL MainThread tlscanary.worker_pool Ctrl-C received. Winding down workers...
2017-07-28 12:29:46 CRITICAL MainThread tlscanary.modes.basemode User abort
```

cr commented:
So far things are looking very good:

```
2017-07-31 12:19:51 INFO Progress tlscanary.progress 0% 800/544372, 0.0% overhead, 799 hosts/min net, 799 hosts/min gross, ETA 2017-07-31 21:37:28
2017-07-31 12:24:52 INFO Progress tlscanary.progress 1% 6650/544372, 0.0% overhead, 1107 hosts/min net, 1107 hosts/min gross, ETA 2017-07-31 20:04:54
2017-07-31 12:29:52 INFO Progress tlscanary.progress 2% 12450/544372, 0.0% overhead, 1131 hosts/min net, 1131 hosts/min gross, ETA 2017-07-31 20:08:59
2017-07-31 12:34:52 INFO Progress tlscanary.progress 3% 18500/544372, 0.0% overhead, 1155 hosts/min net, 1155 hosts/min gross, ETA 2017-07-31 20:00:49
2017-07-31 12:39:53 INFO Progress tlscanary.progress 4% 24300/544372, 0.0% overhead, 1156 hosts/min net, 1156 hosts/min gross, ETA 2017-07-31 20:08:48
2017-07-31 12:44:54 INFO Progress tlscanary.progress 5% 27218/544372, 7.2% overhead, 1045 hosts/min net, 1120 hosts/min gross, ETA 2017-08-01 03:27:58
2017-07-31 12:49:54 INFO Progress tlscanary.progress 6% 30568/544372, 10.9% overhead, 985 hosts/min net, 1092 hosts/min gross, ETA 2017-08-01 01:07:18
2017-07-31 12:54:55 INFO Progress tlscanary.progress 7% 36518/544372, 9.1% overhead, 1013 hosts/min net, 1105 hosts/min gross, ETA 2017-07-31 20:13:15
2017-07-31 12:59:55 INFO Progress tlscanary.progress 8% 42318/544372, 7.9% overhead, 1031 hosts/min net, 1112 hosts/min gross, ETA 2017-07-31 20:13:15
2017-07-31 13:04:55 INFO Progress tlscanary.progress 9% 48018/544372, 6.9% overhead, 1042 hosts/min net, 1115 hosts/min gross, ETA 2017-07-31 20:17:12
2017-07-31 13:09:56 INFO Progress tlscanary.progress 10% 53818/544372, 6.2% overhead, 1054 hosts/min net, 1119 hosts/min gross, ETA 2017-07-31 20:13:20
2017-07-31 13:14:56 INFO Progress tlscanary.progress 10% 54436/544372, 14.2% overhead, 971 hosts/min net, 1109 hosts/min gross, ETA 2017-08-04 09:38:44
2017-07-31 13:19:56 INFO Progress tlscanary.progress 11% 58736/544372, 13.5% overhead, 962 hosts/min net, 1091 hosts/min gross, ETA 2017-07-31 22:21:41
2017-07-31 13:24:57 INFO Progress tlscanary.progress 12% 64536/544372, 12.3% overhead, 976 hosts/min net, 1096 hosts/min gross, ETA 2017-07-31 20:19:09
2017-07-31 13:29:57 INFO Progress tlscanary.progress 13% 70536/544372, 11.2% overhead, 992 hosts/min net, 1103 hosts/min gross, ETA 2017-07-31 20:19:03
2017-07-31 13:34:58 INFO Progress tlscanary.progress 14% 76286/544372, 10.4% overhead, 1002 hosts/min net, 1106 hosts/min gross, ETA 2017-07-31 20:22:42
2017-07-31 13:40:02 INFO Progress tlscanary.progress 15% 81686/544372, 9.7% overhead, 1006 hosts/min net, 1104 hosts/min gross, ETA 2017-07-31 20:21:57
2017-07-31 13:45:02 INFO Progress tlscanary.progress 15% 81854/544372, 15.4% overhead, 950 hosts/min net, 1096 hosts/min gross, ETA 2017-08-08 09:14:25
2017-07-31 13:50:03 INFO Progress tlscanary.progress 16% 87454/544372, 14.4% overhead, 959 hosts/min net, 1098 hosts/min gross, ETA 2017-07-31 20:27:59
2017-07-31 13:55:03 INFO Progress tlscanary.progress 17% 93254/544372, 13.5% overhead, 969 hosts/min net, 1101 hosts/min gross, ETA 2017-07-31 20:24:28
2017-07-31 14:00:04 INFO Progress tlscanary.progress 18% 99004/544372, 12.8% overhead, 978 hosts/min net, 1103 hosts/min gross, ETA 2017-07-31 20:27:57
2017-07-31 14:05:04 INFO Progress tlscanary.progress 19% 104754/544372, 12.1% overhead, 986 hosts/min net, 1105 hosts/min gross, ETA 2017-07-31 20:27:56
2017-07-31 14:10:05 INFO Progress tlscanary.progress 20% 108872/544372, 12.4% overhead, 979 hosts/min net, 1101 hosts/min gross, ETA 2017-07-31 23:20:08
2017-07-31 14:15:06 INFO Progress tlscanary.progress 20% 109672/544372, 16.0% overhead, 944 hosts/min net, 1095 hosts/min gross, ETA 2017-08-02 10:23:15
2017-07-31 14:20:06 INFO Progress tlscanary.progress 21% 115472/544372, 15.2% overhead, 952 hosts/min net, 1097 hosts/min gross, ETA 2017-07-31 20:30:24
2017-07-31 14:25:07 INFO Progress tlscanary.progress 22% 121222/544372, 14.5% overhead, 960 hosts/min net, 1099 hosts/min gross, ETA 2017-07-31 20:33:49
2017-07-31 14:30:07 INFO Progress tlscanary.progress 23% 126922/544372, 13.8% overhead, 967 hosts/min net, 1101 hosts/min gross, ETA 2017-07-31 20:33:47
2017-07-31 14:35:08 INFO Progress tlscanary.progress 24% 132672/544372, 13.2% overhead, 974 hosts/min net, 1103 hosts/min gross, ETA 2017-07-31 20:33:46
2017-07-31 14:40:08 INFO Progress tlscanary.progress 25% 136090/544372, 14.1% overhead, 963 hosts/min net, 1099 hosts/min gross, ETA 2017-08-01 00:50:28
2017-07-31 14:45:08 INFO Progress tlscanary.progress 25% 137290/544372, 16.6% overhead, 939 hosts/min net, 1095 hosts/min gross, ETA 2017-08-01 18:29:18
2017-07-31 14:50:09 INFO Progress tlscanary.progress 26% 143090/544372, 16.0% overhead, 946 hosts/min net, 1097 hosts/min gross, ETA 2017-07-31 20:36:30
2017-07-31 14:55:09 INFO Progress tlscanary.progress 27% 148790/544372, 15.3% overhead, 952 hosts/min net, 1098 hosts/min gross, ETA 2017-07-31 20:42:48
2017-07-31 15:00:10 INFO Progress tlscanary.progress 28% 154590/544372, 14.8% overhead, 958 hosts/min net, 1100 hosts/min gross, ETA 2017-07-31 20:36:38
2017-07-31 15:05:10 INFO Progress tlscanary.progress 29% 160340/544372, 14.2% overhead, 964 hosts/min net, 1101 hosts/min gross, ETA 2017-07-31 20:39:38
2017-07-31 15:10:11 INFO Progress tlscanary.progress 30% 163308/544372, 15.2% overhead, 953 hosts/min net, 1098 hosts/min gross, ETA 2017-08-01 02:03:36
2017-07-31 15:15:11 INFO Progress tlscanary.progress 30% 164308/544372, 17.3% overhead, 932 hosts/min net, 1093 hosts/min gross, ETA 2017-08-01 22:18:26
2017-07-31 15:20:12 INFO Progress tlscanary.progress 31% 170108/544372, 16.7% overhead, 938 hosts/min net, 1095 hosts/min gross, ETA 2017-07-31 20:43:14
2017-07-31 15:25:13 INFO Progress tlscanary.progress 32% 175808/544372, 16.2% overhead, 943 hosts/min net, 1096 hosts/min gross, ETA 2017-07-31 20:49:09
2017-07-31 15:30:13 INFO Progress tlscanary.progress 33% 181608/544372, 15.6% overhead, 949 hosts/min net, 1097 hosts/min gross, ETA 2017-07-31 20:43:20
2017-07-31 15:35:13 INFO Progress tlscanary.progress 34% 187358/544372, 15.2% overhead, 954 hosts/min net, 1099 hosts/min gross, ETA 2017-07-31 20:43:25
2017-07-31 15:40:14 INFO Progress tlscanary.progress 35% 190526/544372, 15.9% overhead, 946 hosts/min net, 1097 hosts/min gross, ETA 2017-08-01 01:28:48
2017-07-31 15:45:15 INFO Progress tlscanary.progress 35% 190726/544372, 18.2% overhead, 924 hosts/min net, 1093 hosts/min gross, ETA 2017-08-06 12:30:29
2017-07-31 15:50:15 INFO Progress tlscanary.progress 36% 196526/544372, 17.7% overhead, 930 hosts/min net, 1094 hosts/min gross, ETA 2017-07-31 20:50:29
2017-07-31 15:55:16 INFO Progress tlscanary.progress 37% 202326/544372, 17.2% overhead, 935 hosts/min net, 1096 hosts/min gross, ETA 2017-07-31 20:50:30
2017-07-31 16:00:16 INFO Progress tlscanary.progress 38% 208026/544372, 16.7% overhead, 940 hosts/min net, 1097 hosts/min gross, ETA 2017-07-31 20:55:54
2017-07-31 16:05:17 INFO Progress tlscanary.progress 39% 213776/544372, 16.3% overhead, 944 hosts/min net, 1098 hosts/min gross, ETA 2017-07-31 20:53:12
2017-07-31 16:10:17 INFO Progress tlscanary.progress 40% 217744/544372, 16.5% overhead, 941 hosts/min net, 1096 hosts/min gross, ETA 2017-07-31 23:16:44
2017-07-31 16:15:18 INFO Progress tlscanary.progress 40% 217744/544372, 18.8% overhead, 921 hosts/min net, 1094 hosts/min gross, ETA --
2017-07-31 16:20:19 INFO Progress tlscanary.progress 41% 222944/544372, 18.3% overhead, 923 hosts/min net, 1093 hosts/min gross, ETA 2017-07-31 21:10:07
2017-07-31 16:25:19 INFO Progress tlscanary.progress 42% 228894/544372, 17.9% overhead, 929 hosts/min net, 1094 hosts/min gross, ETA 2017-07-31 20:57:37
2017-07-31 16:30:20 INFO Progress tlscanary.progress 43% 234594/544372, 17.4% overhead, 933 hosts/min net, 1095 hosts/min gross, ETA 2017-07-31 21:02:36
2017-07-31 16:35:20 INFO Progress tlscanary.progress 44% 240394/544372, 17.0% overhead, 937 hosts/min net, 1097 hosts/min gross, ETA 2017-07-31 20:57:44
2017-07-31 16:40:21 INFO Progress tlscanary.progress 45% 244962/544372, 16.8% overhead, 937 hosts/min net, 1095 hosts/min gross, ETA 2017-07-31 22:09:42
2017-07-31 16:45:21 INFO Progress tlscanary.progress 45% 244962/544372, 19.1% overhead, 919 hosts/min net, 1095 hosts/min gross, ETA --
2017-07-31 16:50:22 INFO Progress tlscanary.progress 46% 250112/544372, 18.8% overhead, 921 hosts/min net, 1094 hosts/min gross, ETA 2017-07-31 21:33:15
2017-07-31 16:55:22 INFO Progress tlscanary.progress 47% 255912/544372, 18.4% overhead, 925 hosts/min net, 1095 hosts/min gross, ETA 2017-07-31 21:04:22
2017-07-31 17:00:23 INFO Progress tlscanary.progress 48% 261612/544372, 18.0% overhead, 929 hosts/min net, 1096 hosts/min gross, ETA 2017-07-31 21:08:54
2017-07-31 17:05:24 INFO Progress tlscanary.progress 49% 267362/544372, 17.6% overhead, 933 hosts/min net, 1097 hosts/min gross, ETA 2017-07-31 21:06:40
2017-07-31 17:10:24 INFO Progress tlscanary.progress 50% 272180/544372, 17.3% overhead, 934 hosts/min net, 1095 hosts/min gross, ETA 2017-07-31 21:57:14
2017-07-31 17:15:25 INFO Progress tlscanary.progress 50% 272180/544372, 19.4% overhead, 918 hosts/min net, 1096 hosts/min gross, ETA --
2017-07-31 17:20:25 INFO Progress tlscanary.progress 51% 276630/544372, 19.3% overhead, 917 hosts/min net, 1094 hosts/min gross, ETA 2017-07-31 22:08:18
2017-07-31 17:25:26 INFO Progress tlscanary.progress 52% 282380/544372, 18.9% overhead, 921 hosts/min net, 1095 hosts/min gross, ETA 2017-07-31 21:13:37
2017-07-31 17:30:26 INFO Progress tlscanary.progress 53% 288130/544372, 18.5% overhead, 925 hosts/min net, 1096 hosts/min gross, ETA 2017-07-31 21:13:36
2017-07-31 17:35:27 INFO Progress tlscanary.progress 54% 293880/544372, 18.1% overhead, 928 hosts/min net, 1096 hosts/min gross, ETA 2017-07-31 21:13:41
2017-07-31 17:40:29 INFO Progress tlscanary.progress 55% 299430/544372, 17.8% overhead, 931 hosts/min net, 1097 hosts/min gross, ETA 2017-07-31 21:16:15
2017-07-31 17:45:28 INFO Progress tlscanary.progress 55% 299398/544372, 19.6% overhead, 917 hosts/min net, 1096 hosts/min gross, ETA --
2017-07-31 17:50:29 INFO Progress tlscanary.progress 56% 303098/544372, 19.7% overhead, 914 hosts/min net, 1094 hosts/min gross, ETA 2017-07-31 23:03:25
2017-07-31 17:55:30 INFO Progress tlscanary.progress 57% 308798/544372, 19.3% overhead, 917 hosts/min net, 1094 hosts/min gross, ETA 2017-07-31 21:22:32
2017-07-31 18:00:31 INFO Progress tlscanary.progress 58% 314548/544372, 19.0% overhead, 921 hosts/min net, 1095 hosts/min gross, ETA 2017-07-31 21:20:45
2017-07-31 18:05:32 INFO Progress tlscanary.progress 59% 320248/544372, 18.6% overhead, 924 hosts/min net, 1096 hosts/min gross, ETA 2017-07-31 21:20:43
2017-07-31 18:10:32 INFO Progress tlscanary.progress 60% 325998/544372, 18.3% overhead, 927 hosts/min net, 1097 hosts/min gross, ETA 2017-07-31 21:19:01
2017-07-31 18:15:32 INFO Progress tlscanary.progress 60% 326616/544372, 19.6% overhead, 916 hosts/min net, 1095 hosts/min gross, ETA 2017-08-02 12:47:56
2017-07-31 18:20:32 INFO Progress tlscanary.progress 61% 329416/544372, 20.1% overhead, 911 hosts/min net, 1093 hosts/min gross, ETA 2017-08-01 00:28:30
2017-07-31 18:25:33 INFO Progress tlscanary.progress 62% 335066/544372, 19.7% overhead, 914 hosts/min net, 1094 hosts/min gross, ETA 2017-07-31 21:31:17
2017-07-31 18:30:34 INFO Progress tlscanary.progress 63% 340816/544372, 19.4% overhead, 917 hosts/min net, 1095 hosts/min gross, ETA 2017-07-31 21:27:51
2017-07-31 18:35:34 INFO Progress tlscanary.progress 64% 346566/544372, 19.1% overhead, 920 hosts/min net, 1095 hosts/min gross, ETA 2017-07-31 21:27:51
2017-07-31 18:40:35 INFO Progress tlscanary.progress 65% 352266/544372, 18.8% overhead, 923 hosts/min net, 1096 hosts/min gross, ETA 2017-07-31 21:27:58
2017-07-31 18:45:36 INFO Progress tlscanary.progress 65% 353834/544372, 19.6% overhead, 915 hosts/min net, 1095 hosts/min gross, ETA 2017-08-01 06:03:09
2017-07-31 18:50:36 INFO Progress tlscanary.progress 65% 355634/544372, 20.5% overhead, 908 hosts/min net, 1094 hosts/min gross, ETA 2017-08-01 03:18:01
2017-07-31 18:55:37 INFO Progress tlscanary.progress 66% 361334/544372, 20.1% overhead, 911 hosts/min net, 1094 hosts/min gross, ETA 2017-07-31 21:36:29
2017-07-31 19:00:38 INFO Progress tlscanary.progress 67% 367084/544372, 19.8% overhead, 914 hosts/min net, 1095 hosts/min gross, ETA 2017-07-31 21:35:03
2017-07-31 19:05:38 INFO Progress tlscanary.progress 68% 372834/544372, 19.5% overhead, 917 hosts/min net, 1095 hosts/min gross, ETA 2017-07-31 21:35:02
2017-07-31 19:10:39 INFO Progress tlscanary.progress 70% 378534/544372, 19.2% overhead, 919 hosts/min net, 1096 hosts/min gross, ETA 2017-07-31 21:36:24
2017-07-31 19:15:39 INFO Progress tlscanary.progress 70% 381052/544372, 19.8% overhead, 914 hosts/min net, 1095 hosts/min gross, ETA 2017-08-01 00:56:19
2017-07-31 19:20:39 INFO Progress tlscanary.progress 70% 381652/544372, 20.8% overhead, 905 hosts/min net, 1093 hosts/min gross, ETA 2017-08-01 17:49:22
2017-07-31 19:25:40 INFO Progress tlscanary.progress 71% 387452/544372, 20.5% overhead, 908 hosts/min net, 1094 hosts/min gross, ETA 2017-07-31 21:41:11
```

cr commented:
I notice some serious instability in ETA estimation that requires fixing, but apart from that it's working nicely.

Maximum memory consumption in the Python process is around 7.5 GBytes, plus ~80 MBytes per Firefox instance. The EC2 instance has 4 CPUs and 16 GBytes of RAM. It is essentially the smallest config that has network speed rated as "High".

cr commented:
So my test run just finished without any issues:

```
2017-07-31 21:25:54 INFO Progress tlscanary.progress 90% 490924/544372, 21.8% overhead, 897 hosts/min net, 1093 hosts/min gross, ETA 2017-08-01 01:41:19
2017-07-31 21:30:55 INFO Progress tlscanary.progress 91% 496674/544372, 21.6% overhead, 900 hosts/min net, 1094 hosts/min gross, ETA 2017-07-31 22:12:28
2017-07-31 21:35:56 INFO Progress tlscanary.progress 92% 502374/544372, 21.3% overhead, 902 hosts/min net, 1094 hosts/min gross, ETA 2017-07-31 22:12:51
2017-07-31 21:40:57 INFO Progress tlscanary.progress 93% 508124/544372, 21.1% overhead, 904 hosts/min net, 1095 hosts/min gross, ETA 2017-07-31 22:12:31
2017-07-31 21:45:57 INFO Progress tlscanary.progress 94% 514024/544372, 20.8% overhead, 906 hosts/min net, 1095 hosts/min gross, ETA 2017-07-31 22:12:23
2017-07-31 21:50:58 INFO Progress tlscanary.progress 95% 517142/544372, 21.0% overhead, 904 hosts/min net, 1094 hosts/min gross, ETA 2017-07-31 22:35:19
2017-07-31 21:55:59 INFO Progress tlscanary.progress 95% 517142/544372, 22.1% overhead, 896 hosts/min net, 1094 hosts/min gross, ETA --
2017-07-31 22:01:00 INFO Progress tlscanary.progress 96% 522542/544372, 21.8% overhead, 898 hosts/min net, 1094 hosts/min gross, ETA 2017-07-31 22:20:52
2017-07-31 22:06:01 INFO Progress tlscanary.progress 97% 528242/544372, 21.6% overhead, 900 hosts/min net, 1094 hosts/min gross, ETA 2017-07-31 22:20:12
2017-07-31 22:11:02 INFO Progress tlscanary.progress 98% 533992/544372, 21.4% overhead, 902 hosts/min net, 1095 hosts/min gross, ETA 2017-07-31 22:20:04
2017-07-31 22:16:02 INFO Progress tlscanary.progress 99% 539692/544372, 21.2% overhead, 904 hosts/min net, 1095 hosts/min gross, ETA 2017-07-31 22:20:09
2017-07-31 22:21:03 INFO Progress tlscanary.progress 100% 544360/544372, 21.0% overhead, 904 hosts/min net, 1094 hosts/min gross, ETA 2017-07-31 22:21:04
```

cr commented:
So far I can't reproduce your stall, @mwobensmith, and I am suspecting that something else may be going on. Can you tell me which EC2 instance size you tested on? Is it possible that you ran into some sort of rate limiting in its tier? Amazon has lately introduced rate limiting on at least disk I/O once you reach a certain volume to encourage upgrading to the next-higher disk tier. There might be something similar going on with network I/O.

I have introduced three changes: Split work into 50 chunks (up from 20) to further reduce memory footprint, synchronize progress reporting to chunks, and tweak progress reporting slightly to be hopefully more accurate when it comes to ETA. I'll startup another full run with that for tonight.

cr commented:
Tonight's full test run went perfectly again. Results are relatively stable apart from the occasional persistent timeout. First ETA was only off by an hour, then stabilized nicely within the first hour, and it was almost spot on since reaching the 50% mark. To be fair, the host data set is not homogeneously distributed and timeout error rate increases towards the end.

```
2017-07-31 23:03:08 INFO MainThread tlscanary.modes.regression Progress: 2% 10887/544372, 10.3% overhead, 970 hosts/min net, 1071 hosts/min gross, ETA 2017-08-01 08:12:26
2017-07-31 23:14:53 INFO MainThread tlscanary.modes.regression Progress: 4% 21774/544372, 11.4% overhead, 947 hosts/min net, 1056 hosts/min gross, ETA 2017-08-01 08:26:21
2017-07-31 23:26:49 INFO MainThread tlscanary.modes.regression Progress: 6% 32661/544372, 12.6% overhead, 896 hosts/min net, 1033 hosts/min gross, ETA 2017-08-01 08:44:12
2017-07-31 23:39:07 INFO MainThread tlscanary.modes.regression Progress: 8% 43548/544372, 13.7% overhead, 861 hosts/min net, 1023 hosts/min gross, ETA 2017-08-01 08:55:41
2017-07-31 23:51:13 INFO MainThread tlscanary.modes.regression Progress: 10% 54435/544372, 14.4% overhead, 847 hosts/min net, 1026 hosts/min gross, ETA 2017-08-01 08:57:40
2017-08-01 00:02:52 INFO MainThread tlscanary.modes.regression Progress: 12% 65322/544372, 14.5% overhead, 861 hosts/min net, 1040 hosts/min gross, ETA 2017-08-01 08:50:15
2017-08-01 00:15:21 INFO MainThread tlscanary.modes.regression Progress: 14% 76209/544372, 15.1% overhead, 851 hosts/min net, 1036 hosts/min gross, ETA 2017-08-01 08:55:19
2017-08-01 00:27:48 INFO MainThread tlscanary.modes.regression Progress: 16% 87096/544372, 15.6% overhead, 840 hosts/min net, 1031 hosts/min gross, ETA 2017-08-01 09:00:16
2017-08-01 00:39:33 INFO MainThread tlscanary.modes.regression Progress: 18% 97983/544372, 15.7% overhead, 834 hosts/min net, 1034 hosts/min gross, ETA 2017-08-01 08:59:15
2017-08-01 00:52:03 INFO MainThread tlscanary.modes.regression Progress: 20% 108870/544372, 16.0% overhead, 833 hosts/min net, 1033 hosts/min gross, ETA 2017-08-01 09:01:16
[...]
2017-08-01 05:52:28 INFO MainThread tlscanary.modes.regression Progress: 68% 370158/544372, 20.7% overhead, 784 hosts/min net, 1056 hosts/min gross, ETA 2017-08-01 09:11:36
2017-08-01 06:05:29 INFO MainThread tlscanary.modes.regression Progress: 70% 381045/544372, 20.8% overhead, 768 hosts/min net, 1039 hosts/min gross, ETA 2017-08-01 09:15:26
2017-08-01 06:18:04 INFO MainThread tlscanary.modes.regression Progress: 72% 391932/544372, 21.0% overhead, 784 hosts/min net, 1057 hosts/min gross, ETA 2017-08-01 09:12:31
2017-08-01 06:30:45 INFO MainThread tlscanary.modes.regression Progress: 74% 402819/544372, 21.0% overhead, 773 hosts/min net, 1045 hosts/min gross, ETA 2017-08-01 09:14:44
2017-08-01 06:43:25 INFO MainThread tlscanary.modes.regression Progress: 76% 413706/544372, 21.2% overhead, 786 hosts/min net, 1061 hosts/min gross, ETA 2017-08-01 09:12:35
2017-08-01 06:56:01 INFO MainThread tlscanary.modes.regression Progress: 78% 424593/544372, 21.3% overhead, 786 hosts/min net, 1060 hosts/min gross, ETA 2017-08-01 09:13:01
2017-08-01 07:08:35 INFO MainThread tlscanary.modes.regression Progress: 80% 435480/544372, 21.4% overhead, 792 hosts/min net, 1068 hosts/min gross, ETA 2017-08-01 09:12:21
2017-08-01 07:21:21 INFO MainThread tlscanary.modes.regression Progress: 82% 446367/544372, 21.5% overhead, 785 hosts/min net, 1059 hosts/min gross, ETA 2017-08-01 09:13:45
2017-08-01 07:34:28 INFO MainThread tlscanary.modes.regression Progress: 84% 457254/544372, 21.5% overhead, 767 hosts/min net, 1040 hosts/min gross, ETA 2017-08-01 09:16:19
2017-08-01 07:47:05 INFO MainThread tlscanary.modes.regression Progress: 86% 468141/544372, 21.6% overhead, 767 hosts/min net, 1042 hosts/min gross, ETA 2017-08-01 09:16:04
2017-08-01 07:59:39 INFO MainThread tlscanary.modes.regression Progress: 88% 479028/544372, 21.7% overhead, 774 hosts/min net, 1051 hosts/min gross, ETA 2017-08-01 09:15:18
2017-08-01 08:12:13 INFO MainThread tlscanary.modes.regression Progress: 90% 489915/544372, 21.9% overhead, 791 hosts/min net, 1078 hosts/min gross, ETA 2017-08-01 09:13:48
2017-08-01 08:24:50 INFO MainThread tlscanary.modes.regression Progress: 92% 500802/544372, 21.9% overhead, 794 hosts/min net, 1079 hosts/min gross, ETA 2017-08-01 09:14:03
2017-08-01 08:37:26 INFO MainThread tlscanary.modes.regression Progress: 94% 511689/544372, 22.0% overhead, 791 hosts/min net, 1083 hosts/min gross, ETA 2017-08-01 09:14:17
2017-08-01 08:50:23 INFO MainThread tlscanary.modes.regression Progress: 96% 522576/544372, 22.1% overhead, 778 hosts/min net, 1063 hosts/min gross, ETA 2017-08-01 09:15:25
2017-08-01 09:03:07 INFO MainThread tlscanary.modes.regression Progress: 98% 533463/544372, 22.2% overhead, 773 hosts/min net, 1067 hosts/min gross, ETA 2017-08-01 09:15:36
2017-08-01 09:15:54 INFO MainThread tlscanary.modes.regression Progress: 100% 544350/544372, 22.3% overhead, 766 hosts/min net, 1061 hosts/min gross, ETA 2017-08-01 09:15:55
```

cr commented:
Just finished a `scan -l 10000` run. All good.

mwobensmith commented:
Latest full regression run with your changes finished just fine. I'm running it again for another test, and if that also passes, I'm confident that we can check this in.

mwobensmith commented:
Run completed without issues, let's land it.

cr commented:
Thanks for the review and all the testing. Will squash and land in a jiffy.

cr commented:
This fixes #112 and concludes #44.
